src/lib/statistics/statistics_objects/distinct_value_count.cpp

```diff
@@ -0,0 +1,18 @@
+#include "distinct_value_count.hpp"
+
+namespace hyrise {
+
+DistinctValueCount::DistinctValueCount(const size_t init_count)
+    : AbstractStatisticsObject(DataType::Long), count{init_count} {}
+
+std::shared_ptr<AbstractStatisticsObject> DistinctValueCount::sliced(
+    const PredicateCondition /* predicate_condition */, const AllTypeVariant& /* variant_value */,
+    const std::optional<AllTypeVariant>& /* variant_value2 */) const {
+  return std::make_shared<DistinctValueCount>(count);
+}
+
+std::shared_ptr<AbstractStatisticsObject> DistinctValueCount::scaled(const Selectivity /* selectivity */) const {
+  return std::make_shared<DistinctValueCount>(count);
```

Comment:
I guess it is not used anywhere right now. But what was your thinking here?

Reply:
Returning the same count or implementing it in general? We could also `Faill("Not implemented")` here. Copying the value rather than doing sth like `std::min(count * selectivity, 1)` was intended due to data skew.

Reply:
Well, we cannot say that the distinct values will get fewer. But at the same time, we cannot say they will stay the same.

Reply:
Neither can we for the `NULL` value ratio, yet we assume so: https://github.com/hyrise/hyrise/blob/3d7fe2ed01b7647c9f54cf1c3760dbc0129b369b/src/lib/statistics/statistics_objects/null_value_ratio_statistics.cpp#L17

However, we also return the old `NULL` value ratio given slicing by a predicate: https://github.com/hyrise/hyrise/blob/3d7fe2ed01b7647c9f54cf1c3760dbc0129b369b/src/lib/statistics/statistics_objects/null_value_ratio_statistics.cpp#L13
From my intuition, this is not correct: If we have any predicate that is neither `IS NULL` nor `IS NOT NULL`, the ratio should be 0. Same for `IS NOT NULL`, and 1 for `IS NULL` if there were `NULL` values before (current ratio > 0) or 0 else.

Reply:
Had a deeper look into `AttributeStatistics` and the `CardinalityEstimator`. `AbstractStatisticsObject::sliced()` is called exactly once here: https://github.com/hyrise/hyrise/blob/34bc2fb8f0eb4ca50f0df9e1431ac4e879a7c4ec/src/lib/statistics/cardinality_estimator.cpp#L747

Note that it is directly called on the histogram object, *not* on an `AttributeStatistics` (container that holds *all* statistics objects). Thus, both `DistinctValueCount` and `NullValueRatioStatistics` can `Fail()` when `sliced()` is called. In fact, the only test that fails when we let `AttributeStatistics::sliced()` `Fail` is the test for this specific method. Unfortunately, that's the only case when we can make a meaningful decision for `NullValueRatioStatistics` (must be 0 when a predicate filters the column). However, this single call only seems to happen because the `CardinalityEstimator` just forwards the histogram for predicates (and not the other statistics).

This is not the case for `AttributeStatistics::scaled()`. This method is applied quite often. However, we cannot make a real assumption about scaled `NullValueRatioStatistics` or `DistinctValueCount` objects.  With these insights, I suggest the following adaptions:

1. Rater than only forwarding the sliced histograms, estimates of predicates should forward the whole sliced `AttributeStatistics`.
2. `DistinctValueCount::sliced()`, `DistinctValueCount::scaled()`, and `NullValueRatioStatistics::scaled()` fail and are *not* called by `AttributeStatistics::scaled()/sliced()`.
3. I check if there are performance differences with `benchmark_all.sh`

Reply:
Hard for me to follow. Let's discuss this in the fortnightly.